TLA+ model for SwarmKit tasks

talex5 opened the pull request:
**- What I did**

I want to learn about how SwarmKit works, and extend its documentation. As a first step, I've been reading through the docs in the `design` directory and writing a more formal version of them. If someone familiar with SwarmKit has time to read over and correct my intepretation, that would be great. I marked with `XXX` things I was particularly unsure about. For example, the docs say very little about how a service is removed.

To make it more precise, I've written the document using [TLA+](https://en.wikipedia.org/wiki/TLA%2B) notation, but hopefully it's still readable to people not familiar with that.

So far, this is based only on reading the design docs. I plan to start reading through the Go code soon, but I wanted to get some early feedback first. Once it's correct, it could be added to the existing documentation.

**- How to test it**

It's easiest to read the [PDF rendering of the spec](https://github.com/docker/swarmkit/files/1931278/Task.pdf).

You can also load the spec into [the TLA toolbox](http://lamport.azurewebsites.net/tla/toolbox.html) and run the model checker on it to test the claimed properties. The document explains how to do that.

**- Description for the changelog**

Add a TLA+ spec for SwarmKit's task model

dperny commented on design/tla/Task.tla:
```diff
@@ -0,0 +1,760 @@
+    (* Every task has a service:
+
+       XXX: The spec says: ``In some cases, there are tasks that exist independent of any service.
+            These do not have a value set in service_id.''. Need an example of one. *)
```
Network attachment tasks. They're an abomination, an absolutely disgusting hack that should never have been shipped. They are tasks which use the `NetworkAttachmentSpec` runtime spec, a spec which contains a container ID. These tasks are created through the grpc server implemented in `resourceapi`, and they are created _by the agent_. The purpose of Network Attachment tasks is to attach regular `docker run` containers to a swarmkit overlay network that has been created with the `--attachable` option.

There have been bugs in the past with attachment tasks because we have no visibility into their lifecycle and their creation and destruction is entirely at the whim of the agent. The end result is that network attachment tasks would remain in the object store after they had been terminated, leaking resources and causing the associated attached networks to be unremovable.

I added a workaround to remove network attachment tasks when a node is removed in #2409, but the underlying issue IIRC remains.

I'm sorry.

dperny replied:
Also, @jlhawn is working on a hack week project to run one-off tasks, like `docker run`-esque tasks, that would exist independently of services.

dperny commented on design/tla/Task.tla:
```diff
@@ -0,0 +1,760 @@
+Transition == {
+  << null, new >>,
+  << new, pending >>,
+  << pending, assigned >>,
```
I can't quite remember, I'd have to go check the code, but it may be possible to move from new to a terminal state, if the task cannot be allocated because there aren't enough free resources. I think. I'm not sure.

dperny commented on design/tla/Task.tla:
```diff
@@ -0,0 +1,760 @@
+  << shutdown, remove >>,
+  << rejected, remove >>,
+  << complete, remove >>,
+  << failed, remove >>,
```
Tasks will never actually BE in the removed state. They will only be in the desired state of removed. The state functions mainly as a flag to the task reaper, to indicate that a task once in a terminal state it and its history can be deleted.

dperny commented on design/tla/Task.tla:
```diff
@@ -0,0 +1,760 @@
+  << shutdown, orphaned >>,
+
+  << remove, null >>,
+  << orphaned, null >>
```
Tasks can be deleted (moved to the null state, I think, in terms of the implementation of this model) from any terminal state. For example, when a task history is deleted, the old tasks will all be in terminal states (like `SHUTDOWN` or `FAILED`) and will be deleted straight from those states. Once a task is in a terminal state, the resources it had in use are definitely free and it is safe to remove it.

dperny commented on design/tla/Task.tla:
```diff
@@ -0,0 +1,760 @@
+=============================================================================
+\* Modification History
+\* Last modified Thu Apr 19 14:55:46 BST 2018 by tal
+\* Created Mon Apr 16 11:31:19 BST 2018 by tal
```
this is a dumb nit, but wouldn't the modification history just be tracked in git? Or does this history get baked into the PDF that a TLA+ file creates?

talex5 commented:
Thanks for the detailed review @dperny! I've made some updates:

[Task.pdf](https://github.com/docker/swarmkit/files/1939176/Task.pdf)

- Explicitly model SwarmKit's view of a node's state (up or down). This is needed because we shouldn't automatically orphan nodes as soon as the node is down. It also makes it easier to write
  properties that depend on node state.

- Note that removing services will be modelled once SwarmKit's new design for this is ready.

- Confirm that slots do not exist as separate objects.

- Confirm that already-terminated terms are removed when their slot is no longer needed.

- `remove` is not actually a possible state of a task, only a desired state. The reaper removes any task that has finished and has this as its desired state.

- Model the allocator rejecting tasks.

- Some minor corrections to the existing markdown docs.

Note: I also made a change to the `tasks` variable, which should make things a bit easier to read as the model now mainly talks about things in terms of tasks rather than task IDs.

I also updated the set of valid transitions to group them by the component that performs them, and to check this.

talex5 commented on design/tla/Task.tla:
```diff
@@ -0,0 +1,760 @@
+  \/ CreateService
+  \/ UpdateService
+
+(* XXX: what about removing a service? (not described in any detail in the design docs) *)
```
OK, I'll wait for @nishanttotla to do that (or submit a design for it) and add it to the model then.

To check I understand the current system:

1. The service is removed from the system.
2. The orchestrator sets the desired state for every task without a service to `remove`.
3. Once a task with no service is shut down, the reaper will remove it.

talex5 commented on design/tla/Task.tla:
```diff
@@ -0,0 +1,760 @@
+(* Try to advance containers towards `desired_state' if we're not there yet.
+
+   XXX: Spec says ``A task will progress through the ACCEPTED, PREPARING, and STARTING states on the way to RUNNING.''
+        I'm assuming the `ready' state occurs in between `preparing' and `starting'. *)
```
I think it's for restarting tasks: the restart manager sets the desired state of the new task to `ready` and only moves it to `running` when the old task has finished shutting down. Perhaps it shouldn't even start shutting down the old task until the new one is ready? Also, the docs say it might wait a bit if there is a restart delay configured.

anshulpundir commented on design/task_model.md:
```diff
@@ -127,8 +127,8 @@ that the requested node has the necessary resources, in the case of global
-progress through the `ACCEPTED`, `PREPARING`, and `STARTING` states on the
-way to `RUNNING`. If a task exits without an error code, it moves to the
+progress through the `ACCEPTED`, `PREPARING`, `READY', and `STARTING` states on
+the way to `RUNNING`. If a task exits without an error code, it moves to the
```
>If a task exits without an error code, it moves to the `COMPLETE` state.

Lets also clarify this by saying that the actual state is changed to COMPLETE by the agent.

talex5 replied:
Actually, it says the agent handles these states on the line above the bit shown by GitHub's diff, so maybe it's OK as it is.

dperny commented:
> OK, I'll wait for @nishanttotla to do that (or submit a design for it) and add it to the model then.
> 
> To check I understand the current system:
> 
> 1. The service is removed from the system.
> 2. The orchestrator sets the desired state for every task without a service to remove.
> 3. Once a task with no service is shut down, the reaper will remove it.

More accurately, the service is removed from the system, the process of which involves setting the desired state of all of its tasks `REMOVE`, and when any task in desired state `REMOVE` enters a terminal state, the taskreaper will delete them.

anshulpundir commented:
>More accurately, the service is removed from the system, the process of which involves setting the desired state of all of its tasks REMOVE, and when any task in desired state REMOVE enters a terminal state, the taskreaper will delete them.

Even more accurately, any task which is in a terminal state or in an unassigned state (i.e. < ASSIGNED),  the taskreaper will delete them.

Also:

>The orchestrator sets the desired state for every task without a service to remove.

Not every task without a service, but just the tasks for the service just deleted.

talex5 commented:
Thanks for the explanation. I've added service removal. I modelled it as flagging the service and all its tasks for removal and then removing it when it had no tasks left (i.e. the proposed design rather than the current one). I added a comment about it in the model.

I also updated the reaper to allow removing tasks that haven't yet been assigned (as explained by @anshulpundir), and added a couple of exceptions to the rule that only one component is in charge of each state to cover that.

[Task.pdf](https://github.com/docker/swarmkit/files/1942925/Task.pdf)

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "tla" git@github.com:talex5/swarmkit.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842353920240
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

talex5 commented:
I've now split the spec into multiple files, and created a separate low-level spec for worker nodes. I also added a Makefile so you can build the PDFs and run the model checker without having to install the TLA toolbox (it uses a Docker image). This could also be used in CI.

talex5 commented:
I think this PR is ready for merging now.

anshulpundir commented on design/tla/WorkerImpl.tla:
```diff
@@ -0,0 +1,320 @@
+
+   ``Moving a task to the Orphaned state is not desirable,
+   because it's the one case where we break the otherwise invariant
+   that the agent sets all states past ASSIGNED.''
```
>that the agent sets all states past ASSIGNED.

Clarify that this is the actual state and not the desired state ?

talex5 commented on design/tla/SwarmKit.tla:
```diff
@@ -0,0 +1,654 @@
+(* The user removes a service.
+
+   Note: Currently, SwarmKit deletes the service from its records immediately.
+   However, this isn't right because we need to wait for the resources to be freed.
```
@dperny mentioned that there are also service-level resources that need to be freed. How about:

> However, this isn't right because we need to wait for service-level resources such as Virtual IPs to be freed.

talex5 commented on design/tla/SwarmKit.tla:
```diff
@@ -0,0 +1,654 @@
+        \E id   \in UnusedId(sid, slot) :
+           tasks' = tasks \union { NewTask(sid, slot, id, running) }
+
+(* Add a task if a slot exists, contains no runnable tasks, and we weren't trying to remove it.
```
I think this is correct. According to https://en.wikipedia.org/wiki/Serial_comma:
> A majority of American style guides mandate use of the serial comma

talex5 commented on design/tla/SwarmKit.tla:
```diff
@@ -0,0 +1,654 @@
+(* Forget about tasks in remove or orphan states.
+
+   Orphaned tasks belong to nodes that we are assuming are lost forever.
+   Or, at the very least, it's likely that a node outage lasting longer than the timeout
```
How about:
> Orphaned tasks belong to nodes that we are assuming are lost forever (or have crashed and will come up with nothing running, which is an equally fine outcome).

I think this is what @dperny meant in https://github.com/docker/swarmkit/pull/2613#discussion_r183125943.

talex5 commented on design/tla/SwarmKit.tla:
```diff
@@ -0,0 +1,654 @@
+
+(* The allocator rejects the task because there aren't enough resources.
+
+   XXX: Not clear whether SwarmKit actually does this.
```
He said he wasn't sure: https://github.com/docker/swarmkit/pull/2613#discussion_r183128393
Shall I remove this then?

anshulpundir replied:
@dperny

anshulpundir replied:
Browsing through the code, I don't see a case where the current allocator fails tasks if there aren't enough resources. @talex5

talex5 replied:
OK, thanks for checking! I've removed this bit now.
